spec-tests/src/test/java/org/candlepin/spec/products/ProductVersioningSpecTest.java

```diff
@@ -0,0 +1,664 @@
+            Products.random().id(id).name(name));
+        assertEquals(prod1.getUuid(), prod2.getUuid());
+    }
+
+    @Test
+    public void shouldCreateTwoDistinctProductInstancesWhenDetailsDiffer() {
+        OwnerDTO owner1 = ownerApi.createOwner(Owners.random());
+        OwnerDTO owner2 = ownerApi.createOwner(Owners.random());
+        String id = StringUtil.random("id");
+        String name = StringUtil.random("name");
+
+        ProductDTO prod1 = ownerProductApi.createProductByOwner(owner1.getKey(),
+            Products.random().id(id).name(name));
+        ProductDTO prod2 = ownerProductApi.createProductByOwner(owner2.getKey(),
+            Products.random().id(id).name(name + "2"));
+        assertNotEquals(prod1.getUuid(), prod2.getUuid());
+    }
+
+    @Test
+    public void shouldCreateANewInstanceWhenMakingChangesToAnExistingInstance() {
+        OwnerDTO owner1 = ownerApi.createOwner(Owners.random());
+        String id = StringUtil.random("id");
+        String name = StringUtil.random("name");
+
+        ProductDTO prod1 = ownerProductApi.createProductByOwner(owner1.getKey(),
+            Products.random().id(id).name(name));
+        assertNotNull(prod1);
+        ProductDTO prod2 = ownerProductApi.updateProductByOwner(owner1.getKey(), prod1.getId(),
+            new ProductDTO().name("new product name"));
+        assertNotNull(prod2);
+        assertNotEquals(prod1.getUuid(), prod2.getUuid());
+
+        List<ProductDTO> prods = ownerProductApi.getProductsByOwner(owner1.getKey(), new ArrayList<>());
+        assertThat(prods).hasSize(1);
+        assertNotEquals(prod1.getUuid(), prods.get(0).getUuid());
+        assertEquals(prod2.getUuid(), prods.get(0).getUuid());
+    }
+
+    @Test
+    public void shouldCreateANewProductInstanceWhenAnOrgUpdatesASharedInstance() {
+        OwnerDTO owner1 = ownerApi.createOwner(Owners.random());
+        OwnerDTO owner2 = ownerApi.createOwner(Owners.random());
+        OwnerDTO owner3 = ownerApi.createOwner(Owners.random());
+        String id = StringUtil.random("id");
+        String name = StringUtil.random("name");
+
+        ProductDTO prod1 = ownerProductApi.createProductByOwner(owner1.getKey(),
+            Products.random().id(id).name(name));
+        ProductDTO prod2 = ownerProductApi.createProductByOwner(owner2.getKey(),
+            Products.random().id(id).name(name));
+        ProductDTO prod3 = ownerProductApi.createProductByOwner(owner3.getKey(),
+            Products.random().id(id).name(name));
+        assertEquals(prod1.getUuid(), prod2.getUuid());
+        assertEquals(prod1.getUuid(), prod3.getUuid());
+
+        ProductDTO prod4 = ownerProductApi.updateProductByOwner(owner2.getKey(), id,
+            new ProductDTO().name("new product name"));
+        assertNotEquals(prod4.getUuid(), prod1.getUuid());
+
+        List<ProductDTO> prods = ownerProductApi.getProductsByOwner(owner2.getKey(), new ArrayList<>());
+        assertThat(prods).hasSize(1);
```

Comment:
These can be turned into assertion chains:

```java
assertThat(prods)
  .isNotNull()
  .singleElement()
  .returns(prod4.getUuid(), ProductDTO::getUuid)
  .doesNotReturn(prod2.getUuid(), ProductDTO::getUuid);
```

`.singleElement` checks that the object is a collection that contains exactly one element, but then also extracts that one element for further evaluation.

Alternatively, you can use the `extracting` method to get the UUID before doing normal equals comparisons:

```java
assertThat(prods)
  .isNotNull()
  .singleElement()
  .extracting(ProductDTO::getUuid)
  .isEqualTo(prod4.getUuid())
  .isNotEqualTo(prod2.getUuid());
```